### ISSUE TITLE
Limited functionality for fine-tuning TF models using the TensorFlowTransform

### ISSUE THREAD

**Issue by abgoswam** [dotnet/machinelearning#2348]
The TensorFlowTransform in ML.NET  provides some basic capabilities for re-training pre-trained TF models.  

The table below lists down the capabilities currently supported for fine-tuning TF models. It also lists  functionality that's currently not supported.

 | Fine-Tuning abilities | Supported by ML.NET TensorFlowTransform | 
 | ---------- | -------- | 
 | Continue retraining on TF SavedModel,  given dataset  with same shape as  original model | Yes | 
 | Continue retraining on TF SavedModel,  given dataset  with _different_ shape as  original model | No | 
 | Freeze weights of all layers, except the last layer. Replace last layer of TF SavedModel  with new layer  based on categories of new dataset. Re-train the weights for last layer only.  | No |
 | Freeze weights upto any  arbitrary Nth layer  (given layer name). Replace final layer of TF SavedModel  with new layer  based on categories of new dataset.  Re-train the weights from (N+1) th layer to the last layer.  | No | 

@zeahmed  @JRAlexander

**Comment by abgoswam** [comment]
Some notes (from 9/2018) when we were exploring enabling fine-tuning capabilities for TensorFlow in ML.NET .  It lists some blocking issues we hit at that time : 

1. Gradient Support incomplete in C API
    - Gradients are currently being ported from python to C++ . Several operations have not been ported yet
    - In particular gradient for Select is absent, which blocks some simple models from being re-trained

2. Save&Restore of TF models
    - Currently ML.NET supports the following
        Scoring using Frozen models format
        Scoring using SavedModel format -- model is not mutated
         Retrain using SavedModel format -- model structure is kept as-is.  (This allows us to use the save/Const , save/control_dependency trick)
    - We need to find a generic way of creating a SavedModel  using C API . What we want is retrain using SavedModel format, where model structure can be modified e.g. replace the final layer

Related links:
    - https://github.com/migueldeicaza/TensorFlowSharp/issues/25
    - https://github.com/migueldeicaza/TensorFlowSharp/issues/264
    - https://github.com/tensorflow/tensorflow/tree/master/tensorflow/cc/gradients/
